### PR TITLE
Limit Refuel to 1 GAS

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
@shargon @roman-khimov What do you think?

**Comment by shargon** [comment]
I think that if the account it's related to any transaction in the block, and it remove the total gas can poduce a Denial of service during onPersist, I agree with @roman-khimov in `ABORT`

**Comment by ZhangTao1596** [comment]
Attackers can call refuel multi times to occupy system resource.

**Comment by ZhangTao1596** [comment]
Limit sum of refueled gas to 1 gas?

**Comment by shargon** [comment]
Limit Refuel to only one call per invocation

File: src/neo/SmartContract/ApplicationEngine.cs [hunk 1]
```diff
@@ -160,6 +162,9 @@ internal void Refuel(long gas)
         {
             checked
             {
+                gas_refuel += gas;
+                if (gas_refuel > MAX_REFUEL)
+                    throw new InvalidOperationException();
```
Comment (shargon) [inline src/neo/SmartContract/ApplicationEngine.cs:167]: Add message?

File: src/neo/SmartContract/Native/GasToken.cs [hunk 2]
```diff
@@ -40,7 +40,7 @@ internal override async ContractTask OnPersist(ApplicationEngine engine)
         [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States | CallFlags.AllowNotify)]
         private async ContractTask Refuel(ApplicationEngine engine, UInt160 account, long amount)
         {
-            if (amount < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (amount <= 0) throw new ArgumentOutOfRangeException(nameof(amount));
             if (!engine.CheckWitnessInternal(account)) throw new InvalidOperationException();
             await Burn(engine, account, amount);
             engine.Refuel(amount);
```
Comment (ZhangTao1596) [inline src/neo/SmartContract/Native/GasToken.cs:46]:
```suggestion
            engine.Refuel(amount);
            await Burn(engine, account, amount);
```

Comment (erikzhang) [reply src/neo/SmartContract/Native/GasToken.cs:46]: What's the difference?

Comment (ZhangTao1596) [reply src/neo/SmartContract/Native/GasToken.cs:46]: We don't need do burn when refuel check failed.

**Comment by erikzhang** [comment]
@roman-khimov What do you think?

**Comment by erikzhang** [comment]
Can anyone fix the UT?

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
@superboyiii could you test it if can produce a DoS if during `onPersist`, we remove the gas from other transaction fee?

**Comment by roman-khimov** [comment]
> @roman-khimov What do you think?

1. It's very limiting.
2. It doesn't solve the problem.

`ABORT` is still there to make node's computational efforts almost free, so amplification is still possible.

**Comment by erikzhang** [comment]
Another protection mechanism can also be introduced: we can only allow `Refuel` calls from contracts. In this way, if the function is abused, we can sanction the contracts in the future.

**Comment by erikzhang** [comment]
> Another protection mechanism can also be introduced: we can only allow Refuel calls from contracts. In this way, if the function is abused, we can sanction the contracts in the future.

Or can we even require the contract to pay for Refuel in advance?

**Comment by superboyiii** [comment]
> > @roman-khimov What do you think?
> 
> 1. It's very limiting.
> 2. It doesn't solve the problem.
> 
> `ABORT` is still there to make node's computational efforts almost free, so amplification is still possible.

I agree with Roman. `ABORT` now and think for better solution later. It's too hurry find another way which might cause other issue.

**Comment by shargon** [comment]
> Or can we even require the contract to pay for Refuel in advance?

Too much better

**Comment by chenzhitong** [comment]
Pay GAS in advance and **refund** the remaining system fee to the user. (No refund if an exception occurs)

**Comment by erikzhang** [comment]
I have another idea. We can limit it to `tx.SystemFee + tx.NetworkFee`. What do you think?

**Comment by roman-khimov** [comment]
1. Even more limiting, making Refuel less and less practical.
2. Limits amplification factor to 2. Not a lot, true, but still amplification.

**Comment by erikzhang** [comment]
> Even more limiting, making Refuel less and less practical.

I think it's better than removing it.

> Limits amplification factor to 2. Not a lot, true, but still amplification.

If we limit it to `tx.SystemFee + tx.NetworkFee`, I'm sure no one will attack this way again.

**Comment by shargon** [comment]
Pay gas in advance, make a bag, and decrement this bag during refuel, like this never can affect to OnPersist

File: src/neo/SmartContract/Native/GasToken.cs [hunk 3]
```diff
@@ -40,10 +40,12 @@ internal override async ContractTask OnPersist(ApplicationEngine engine)
         [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States | CallFlags.AllowNotify)]
         private async ContractTask Refuel(ApplicationEngine engine, UInt160 account, long amount)
         {
-            if (amount < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (amount <= 0) throw new ArgumentOutOfRangeException(nameof(amount));
+            if (ContractManagement.GetContract(engine.Snapshot, engine.CallingScriptHash) is null)
```
Comment (shargon) [inline src/neo/SmartContract/Native/GasToken.cs:44]: Account?

Comment (erikzhang) [reply src/neo/SmartContract/Native/GasToken.cs:44]: What do you mean?

**Comment by erikzhang** [comment]
I think it's perfect now.

**Comment by chenzhitong** [comment]
UT

**Comment by roman-khimov** [comment]
> > Even more limiting, making Refuel less and less practical.
> 
> I think it's better than removing it.

The intention behind #2444 can be traced via #2443 to neo-project/proposals#137 and that intention to me is to have contract-sponsored actions. Limiting Refuel to fee amount makes the user inevitably pay half of the real price, so it's a discount, but the intention was rather to have contract paying 90%+ of fees. I've questioned usefulness of this method already in #2443 (from a bit different perspective though), but with this limit it's hard for me to imagine who is going to use it. Methods should solve some problems and I'm not sure it solves anything now.

**Comment by erikzhang** [comment]
It is difficult for a contract to determine whether the sponsorship has been abused. Now it does not need to worry too much, because users have to pay at least half of the cost.

**Comment by erikzhang** [comment]
> but the intention was rather to have contract paying 90%+ of fees

Maybe we can limit it to `10 * (tx.SystemFee + tx.NetworkFee)`?

**Comment by superboyiii** [comment]
> > but the intention was rather to have contract paying 90%+ of fees
> 
> Maybe we can limit it to `10 * (tx.SystemFee + tx.NetworkFee)`?

I think system_fee + = Refuel * exe_Factor/3 is more resonable. Don't make a static limit to a dynamic result.

**Comment by roman-khimov** [comment]
> Maybe we can limit it to `10 * (tx.SystemFee + tx.NetworkFee)`?

Increases usefulness, raises amplification factor.

> Don't make a static limit to a dynamic result.

`ExecFeeFactor` is already priced-in with system fee.

**Comment by roman-khimov** [comment]
> Pay gas in advance, make a bag, and decrement this bag during refuel, like this never can affect to OnPersist

And BTW, while this probably is the only valid solution (if not counting `Refuel` deletion) it would also mean that execution will produce some side-effect for _failed_ transaction which was never the case before (fees written off is a bit different story).

**Comment by roman-khimov** [comment]
As a side note, think of `MaxBlockSystemFee` also that `Refuel` technically breaks. Not the biggest problem with it, but still.

**Comment by ZhangTao1596** [comment]
We should revert first. Seems we can't get agreement soon.

**Comment by igormcoelho** [comment]
For this Refuel to work we need to make sure that the assets are innevitably taken from user or contract account. It should have similar logic to network or system fees, so relating a factor into it is a clever solution.

For the moment, I agree with @erikzhang in putting a limiting factor of 2, or even zero, if necessary. Best thing if we could put it in some global constant, that can be updated on time. Then, we take the time to discuss it further. 

I also agree to take funds only from a deployed contract. 
In my opinion, nodes can calculate how much refunds are being done from a contract, and if it reaches some threshold, Refuel gets disabled for that contract. It means that contracts willing to sponsor anything should be careful to only authorize only effective transactions.

**Comment by ZhangTao1596** [comment]
Another problem. Once we set the limit in`ApplicationEngine`, we have to keep it there and add height check when we apply new refuel mechanism in the future. Because old tx can't execute in new mechanism.

**Pull request closed without merging**